# DSML-Python Refresher 2

**Agenda**
1. Control Flow
2. Functions
3. Lists
4. Strings
5. Tuples
6. Sets
7. Dictionaries
8. Linked List
9. Loops
10. Conditional statements

#**`Control Flow`**

In [ ]:
x = 60

In [ ]:
if x < 50:
    print("random")
    y = 10
    y += 5
    print(y)

print("After the if block!")

After the if block!


In [ ]:
print("0")

if 50 > 40:
    print("1")
    print("2")
    print("3")

print("4")

0
1
2
3
4


In [ ]:
print("0")

if 50 < 40:
    print("1")
    print("2")
    print("3")

print("4")

0
4


In [ ]:
'''
LINES BEFORE THE IF-ELSE BLOCK WILL ALWAYS EXECUTE.

if <condition>:
    IF BLOCK EXECUTES IF THE CONDITION IS TRUE.
    SECOND LINE.
    THIRD LINE.
else:
    ELSE BLOCK EXECUTES IF THE CONDITION IS FALSE.
    SECOND LINE.
    THIRD LINE.

LINES AFTER THE IF-ELSE BLOCK WILL ALWAYS EXECUTE.
'''

'\nLINES BEFORE THE IF-ELSE BLOCK WILL ALWAYS EXECUTE.\n\nif <condition>:\n    IF BLOCK EXECUTES IF THE CONDITION IS TRUE.\n    SECOND LINE.\n    THIRD LINE.\nelse:\n    ELSE BLOCK EXECUTES IF THE CONDITION IS FALSE.\n    SECOND LINE.\n    THIRD LINE.\n\nLINES AFTER THE IF-ELSE BLOCK WILL ALWAYS EXECUTE.\n'

In [ ]:
'''
Question 1 -
1. Take an integer as input from the user.
2. Calculate cube of the input.
3. If result > 50 print "Massive Number" else print "Small Number".
4. After this, print "EXECUTION COMPLETE!"
'''

'\nQuestion 1 - \n1. Take an integer as input from the user.\n2. Calculate cube of the input.\n3. If result > 50 print "Massive Number" else print "Small Number".\n4. After this, print "EXECUTION COMPLETE!"\n'

In [ ]:
n = int(input())

result = n ** 3

if result > 50:
    print("Massive Number!")
else:
    print("Small Number!")

print("EXECUTION COMPLETE!")

3
Small Number!
EXECUTION COMPLETE!


In [ ]:
'''
Question 2 -
1. Take a password as input from User.
2. Take another input to verify password.
3. Print "Successfully Verified!" if both passwords are same.
4. Print "Verification Failed!" otherwise.
'''

'\nQuestion 2 -\n1. Take a password as input from User.\n2. Take another input to verify password.\n3. Print "Successfully Verified!" if both passwords are same.\n4. Print "Verification Failed!" otherwise.\n'

In [ ]:
password_1 = input("Enter your password - ")
password_2 = input("Enter your password again - ")

if password_1 == password_2:
    print("Successfully Verified!")
else:
    print("Verification Failed!")

Enter your password - password
Enter your password again - random123
Verification Failed!


In [ ]:
time = int(input())

if time > 9 and time <= 12:
    print("Good Morning :)")
else:
    if time > 12 and time <= 17:
        print("Good Afternoon :|")
    else:
        if time > 17 and time <= 20:
            print("Good Evening :(")
        else:
            print("Good Night >:(")

14
Good Afternoon :|


In [ ]:
time = int(input())

if time > 9 and time <= 12:
    print("Good Morning :)")
elif time > 12 and time <= 17:
    print("Good Afternoon :)")
elif time > 17 and time <= 20:
    print("Good Evening :)")
else:
    print("Good Night :)")

14
Good Afternoon :)


### Pass statement
\
pass in Python is a null statement which the interpreter simply 'moves over' while executing the code. It can be used in a variety of scenarios where you want to leave parts of your code undefined (such as if-else, loops, classes or methods) but the Python syntax doesn't allow it.

In [ ]:
x = 10

if x % 4 == 0:


IndentationError: expected an indented block (387309564.py, line 4)

In [ ]:
x = 10

if x % 4 == 0:
    pass

In [ ]:
for i in range(10):


IndentationError: expected an indented block (4155345654.py, line 2)

In [ ]:
for i in range(10):
    pass

#**`Break Statement`**
Stop / Interrupt / Destroy the loop. Works with both for and while loop.


**Usage:** The `break` statement is utilized within a loop, often accompanied by a conditional statement (such as an `if` statement) that checks a specific condition.

**Purpose:** When the `break` statement is executed, the loop is immediately terminated, and the program proceeds with the next statement after the loop.

**Example:**
```python
for i in range(5):
    if i == 3:
        break
```

In this example, the loop stops when `i` equals 3, and the remaining numbers (4 and 5) are not iterated over. The `break` statement is a valuable tool for controlling the flow of loops and prematurely exiting a loop when a certain condition is met.

In [ ]:
for i in range(10):
    print(i)

    if i == 5:
        break

    print("After the if")

print("\nOutside the loop")

0
After the if
1
After the if
2
After the if
3
After the if
4
After the if
5

Outside the loop


#**`Continue Statement`**

Skips an iteration.
Works with both for loops and while loops

**Usage:** The `continue` statement is employed within a loop, typically within an `if` statement that checks a specific condition.

**Purpose:** When the `continue` statement is executed, the current iteration is prematurely terminated, and the loop immediately proceeds to the next iteration.

**Example:**
```python
for i in range(5):
    if i == 3:
        continue
```

In this example, when `i` equals 3, the current iteration is skipped, and the loop continues with the next value of `i` (4 and 5). The `continue` statement is a useful tool for controlling the flow of loops and skipping specific iterations based on a condition.

In [ ]:
for i in range(1, 10):
    if i == 7:
        continue

    print(i, end=' ')

1 2 3 4 5 6 8 9 

#### Question
- Write a program that continuously asks the user to provide an input number.
- Print the square of that input number.
- The program should stop if the user enters 5.
- In case user enters any multiple of 4, skip it.

In [ ]:
while True:
    x = int(input())

    if x == 5:
        break

    if x % 4 == 0:
        continue

    print(x ** 2)

#**`Nested Loops`**

**Usage:** Nested loops involve placing one loop inside another, providing a powerful way to iterate through complex data structures, such as multi-dimensional lists or matrices.

**Purpose:** Nested loops are commonly used for tasks like traversing rows and columns in a 2D array or searching for elements in a grid.

**Example:**
```python
for i in range(3):
    for j in range(2):
        print(i, j)
```
In this example, two nested `for` loops are used. The outer loop iterates three times (i = 0, 1, 2), and for each iteration of the outer loop, the inner loop iterates twice (j = 0, 1). This results in a total of six combinations of `i` and `j` being printed.

In [ ]:
for j in range(5):
    print("*", end='')

*****

In [ ]:
for j in range(5):
    print("*", end='')

print()

for j in range(5):
    print("*", end='')

print()

for j in range(5):
    print("*", end='')

*****
*****
*****

In [ ]:
for i in range(3):
    for j in range(5):
        print("*", end='')
    print()

*****
*****
*****


##### **Question**: Write a program to print a `N x M` matrix of `*`

In [ ]:
N = 4 # no of rows
M = 2 # no of cols

**
**
**
**

In [ ]:
N = int(input())
M = int(input())

for i in range(N):
    for j in range(M):
        print('*', end='')
    print()

6
10
**********
**********
**********
**********
**********
**********


In [ ]:
M = 2
for j in range(M):
    print('*', end='')

**

In [ ]:
str1 = "akash"
str1[::-1]

'hsaka'

In [ ]:
for i in range (len(str1)-1, -1, -1):
    print(str1[i])
#     print(i)

h
s
a
k
a


In [ ]:
for i in range(0,-20, -1):
    print(i)

0
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14
-15
-16
-17
-18
-19


## Stair Pattern

In [ ]:
N = 5

for i in range(1, N+1):
    for j in range(i):
        print("*", end='')
    print()

*
**
***
****
*****


## Reverse Stair Pattern

In [ ]:
N = 15

for i in range(1, N+1):
    # For each row, print (N-i+1) cols
    for j in range(N-i+1):
        print('*', end='')
    print()


***************
**************
*************
************
***********
**********
*********
********
*******
******
*****
****
***
**
*


# GCD
Given two integer inputs, write a program to calculate GCD of these two numbers.


In [ ]:
A = 200
B = 80

# Smallest = 1
# Largest = min(a, b)
# I am looking for the highest common factor

for i in range(min(A, B), 0, -1):
    # We wish to find a common factor of A & B
    # Check if (i is a factor of A) and (i is a factor of B)
    if (A % i == 0) and (B % i == 0):
        gcd = i
        break

print(gcd)

40


In [ ]:
min(10, 20)

10

In [ ]:
min(10, 10)

10

In [ ]:
min(30, 24)

24

In [ ]:
max(10, 20)

20


#**`Python Functions`**

Functions in Python are fundamental for organizing and structuring your code. They serve as reusable blocks of code designed to execute specific tasks or computations. Let's explore the key aspects of functions:

- **Function Definition**: Functions are defined using the `def` keyword, followed by the function name, a pair of parentheses, and a colon. The code block that follows is the function body, which contains the instructions to be executed when the function is called.

  Example:
  ```python
  def greet(name):
      print(f"Hello, {name}!")
  ```
  In this example, `greet` is the function name, and it takes a single argument, `name`.

- **Function Call**: To execute a function, you call it by using its name followed by a pair of parentheses. You can pass arguments to the function within the parentheses.

  Example:
  ```python
  greet("Alice")
  ```
  Calling the `greet` function with the argument "Alice" will print "Hello, Alice!" to the console.

- **Return Statement**: Functions can return values using the `return` statement. The returned value can be used in other parts of your program.

  Example:
  ```python
  def add(a, b):
      result = a + b
      return result
  ```

- **Function Parameters**: Functions can accept one or more parameters (also known as arguments) that provide input values to the function. These parameters are defined within the parentheses in the function header.

- **Function Signature**: The function signature is a combination of the function name and the number and types of parameters it accepts. It's a way to uniquely identify a function.

- **Function Reusability**: One of the primary benefits of functions is reusability. You can call the same function multiple times with different arguments, making your code more efficient and easier to maintain.


In [ ]:
# FUNCTION DEFINITION -> "def" is the keyword

def prepare_for_guests():
    print("Dust all the rooms!")
    print("Arrange the living area!")
    print("Go out and get cold drinks and chips!")
    print("Take out the fancy dinner set!")
    print("Touch their feet when you see them!")

In [ ]:
prepare_for_guests()

Dust all the rooms!
Arrange the living area!
Go out and get cold drinks and chips!
Take out the fancy dinner set!
Touch their feet when you see them!


In [ ]:
def sheldon_knock():
    print("knock knock knock Penny!")
    print("knock knock knock Penny!")
    print("knock knock knock Penny!")

In [ ]:
sheldon_knock()

knock knock knock Penny!
knock knock knock Penny!
knock knock knock Penny!


In [ ]:
# FUNCTIONS CAN ACCEPT ARGUMENTS!

def sheldon_knock(name):
    print("knock knock knock", name)
    print("knock knock knock", name)
    print("knock knock knock", name)

In [ ]:
sheldon_knock("Swapnil")

knock knock knock Swapnil
knock knock knock Swapnil
knock knock knock Swapnil


In [ ]:
sheldon_knock("Deepak")

knock knock knock Deepak
knock knock knock Deepak
knock knock knock Deepak


In [ ]:
student = "Vijaylaxmi"

In [ ]:
sheldon_knock(student)

knock knock knock Vijaylaxmi
knock knock knock Vijaylaxmi
knock knock knock Vijaylaxmi


In [ ]:
a = None
b = "Random"

In [ ]:
a = b

In [ ]:
a

'Random'

In [ ]:
user_input = input()

Akash


In [ ]:
user_input

'Akash'

In [ ]:
sheldon_knock(user_input)

knock knock knock Akash
knock knock knock Akash
knock knock knock Akash


In [ ]:
# A FUNCTION CAN ACCEPT MULTIPLE ARGUMENTS!

In [ ]:
def introduce_family(father, mother, sibling, myself):
    print("Father's Name -", father)
    print("Mother's Name -", mother)
    print("Sibling's Name -", sibling)
    print("My Name -", myself)

In [ ]:
introduce_family("Brad Pitt", "Angelina Jolie", "Stormi", "Bipin Kalra")

Father's Name - Brad Pitt
Mother's Name - Angelina Jolie
Sibling's Name - Stormi
My Name - Bipin Kalra


In [ ]:
introduce_family("Angelina Jolie", "Brad Pitt", "Stormi", "Bipin Kalra")

Father's Name - Angelina Jolie
Mother's Name - Brad Pitt
Sibling's Name - Stormi
My Name - Bipin Kalra


In [ ]:
introduce_family("A", "B", "C")

TypeError: introduce_family() missing 1 required positional argument: 'myself'

In [ ]:
introduce_family("A", "B", "C", "D", "E")

TypeError: introduce_family() takes 4 positional arguments but 5 were given

In [ ]:
s1 = 'this string uses single quotes!'

In [ ]:
s2 = "this string uses double quotes!"

In [ ]:
s3 = '''this is a special string that uses triple single quotes! -> DOCSTRING'''

In [ ]:
'''
This string
            maintains white spaces


            which
        include tabs, spaces and newline characters!
'''

'\nThis string\n            maintains white spaces\n            \n            \n            which\n        include tabs, spaces and newline characters!\n'

In [ ]:
def random(a, b):
    '''
    This function takes two arguments -
    a and b can be of any type

    Output - Prints both the arguments separated by a space.
    '''

    print(a, b)

In [ ]:
help(random)

Help on function random in module __main__:

random(a, b)
    This function takes two arguments -
    a and b can be of any type
    
    Output - Prints both the arguments separated by a space.



In [ ]:
# help(print)

In [ ]:
4 + 5

9

In [ ]:
"string1" + "string2"

'string1string2'

In [ ]:
"string" + 1

TypeError: can only concatenate str (not "int") to str

In [ ]:
def add_two(a, b):
    '''
    Arguments -
    a: Number or String
    b: Number or String

    Note - Please do not enter mixed data types i.e. both a and b should be of the same type.

    Output -
    String + String -> Concatenation
    Number + Number -> Addition
    '''

    result = a + b
    print(result)

In [ ]:
help(add_two)

Help on function add_two in module __main__:

add_two(a, b)
    Arguments - 
    a: Number or String
    b: Number or String
    
    Note - Please do not enter mixed data types i.e. both a and b should be of the same type.
    
    Output - 
    String + String -> Concatenation
    Number + Number -> Addition



In [ ]:
def add_two_numbers(a, b):
    print(a + b)

In [ ]:
add_two_numbers(10, 5)

15


In [ ]:
result = add_two_numbers(10, 5)

15


In [ ]:
print(result)

None


In [ ]:
def add_two_numbers(a, b):
    return a + b

In [ ]:
result = add_two_numbers(10, 5)

In [ ]:
print(result)

15


In [ ]:
def add_two_numbers_with_print(a, b):
    print(a + b)

def add_two_numbers_with_return(a, b):
    return a + b

In [ ]:
add_two_numbers_with_print(10, 5)

15


In [ ]:
add_two_numbers_with_return(10, 5)

15

In [ ]:
def add_two_numbers_with_both(a, b):
    value = a + b

    print(value)
    return value

In [ ]:
result = add_two_numbers_with_both(10, 5)

15


In [ ]:
print(result)

15


In [ ]:
def random():
    print("CODE BEFORE RETURN!")
    return 1
    print("CODE AFTER RETURN!")

In [ ]:
random()

CODE BEFORE RETURN!


1

In [ ]:
# FUNCTION DEFINITION
def function_name():
    x = 10
    print("Inside function!")
    print(x)

In [ ]:
# FUNCTION CALL / EXECUTION
function_name()

Inside function!
10


In [ ]:
def random1():
    return True

def random2():
    print(True)

In [ ]:
random1()

True

In [ ]:
random2()

True


In [ ]:
result1 = random1()
result2 = random2()

True


In [ ]:
print(result1, result2, sep = "\n")

True
None


In [ ]:
def simple_interest(p, r, t):
    interest = (p * r * t) / 100

    return interest

In [ ]:
simple_interest(10000, 5, 5)

2500.0

In [ ]:
def simple_interest(p, r, t = 10):
    interest = (p * r * t) / 100

    return interest

In [ ]:
simple_interest(10000, 5)

5000.0

In [ ]:
simple_interest(10000, 5, 20)

10000.0

In [ ]:
def simple_interest(p, r = 5, t):
    interest = (p * r * t) / 100

    return interest

'''
Python does not allow the creation of functions
where there are optional arguments before required arguments!
The order should always be -> REQUIRED followed by OPTIONAL (DEFAULT)
'''

SyntaxError: non-default argument follows default argument (21423845.py, line 1)

In [ ]:
def simple_interest(p, r = 5, t = 10):
    interest = (p * r * t) / 100

    return interest

In [ ]:
simple_interest(10000)

5000.0

In [ ]:
simple_interest(10000, 10)

10000.0

In [ ]:
simple_interest(10000, 10, 20)

20000.0

In [ ]:
a = 10 # global variable

def random1():
    print("Inside random1 -", a)

def random2():
    a = 20 # local variable
    print("Inside random2 -", a)

random1()
random2()

print("Outside -", a)

Inside random1 - 10
Inside random2 - 20
Outside - 10


**Lambda Functions**

Lambda functions, also known as anonymous functions, provide a concise way to define small, single-expression functions without using the `def` keyword. They are often used for simple operations or when you need a quick function for a short task.

**Syntax:** Lambda functions are defined using the `lambda` keyword, followed by the input parameters, a colon, and the expression to be evaluated.

Example:
```python
square = lambda x: x**2
```

**Usage:** Lambda functions are often used when you need a function for a short-lived task, like sorting a list or performing a simple operation on elements in a list.

Example:
```python
numbers = [1, 2, 3, 4, 5]
squared_numbers = list(map(lambda x: x**2, numbers))
```

**Advantages:** Lambda functions are concise and can be defined in a single line of code. They are particularly useful in functional programming and as arguments for functions like `map()`, `filter()`, and `sort()`.

**Limitations:** Lambda functions are restricted to simple expressions and cannot contain multiple statements or complex logic.

In [ ]:
def random1(x):
    return x + 10

# A function that only contains a single statement which is a return value

In [ ]:
random2 = lambda x : x + 10

In [ ]:
random1(5)

15

In [ ]:
random2(5)

15

In [ ]:
hello = lambda : "HELLO" # Lambda function without arguments

In [ ]:
hello()

'HELLO'

In [ ]:
sum_2 = lambda x, y : x + y # Multiple Arguments

In [ ]:
sum_2(10, 5)

15

In [ ]:
simple_interest = lambda p, r = 5, t = 10 : (p * r * t) / 100

In [ ]:
simple_interest(5000)

2500.0

In [ ]:
# Lambda functions can be anonymous

In [ ]:
random = lambda x : x + 10

In [ ]:
random(5)

15

In [ ]:
(lambda x : x + 10)(5) # Anonymous - do not need a name to be called!

15

In [ ]:
import math

In [ ]:
r = 5

In [ ]:
area = math.pi * (r ** 2)

In [ ]:
area

78.53981633974483

In [ ]:
math.pi

3.141592653589793

In [ ]:
math.ceil(5.6)

6

#**`Python Lists`**

A list is a versatile data structure in Python, used for storing ordered and mutable elements.

| List Operation  | Description                                                         | Example                                        | Output               |
|-----------------|---------------------------------------------------------------------|------------------------------------------------|-----------------------|
| **append()**    | Used to add an element to the end of the list.                      | `my_list = [1, 2, 3]`<br>`my_list.append(4)`    | `[1, 2, 3, 4]`       |
| **pop()**       | Used to remove and return the last element of the list.             | `my_list = [1, 2, 3]`<br>`last_element = my_list.pop()` | `last_element = 3`  |
| **len()**       | Used to get the length of the list.                                 | `my_list = [1, 2, 3]`<br>`length = len(my_list)`  | `length = 3`         |
| **sort()**      | Used to sort the list in ascending order.                            | `my_list = [3, 1, 2]`<br>`my_list.sort()`          | `[1, 2, 3]`          |

In [ ]:
# CRUD -> CREATE, READ, UPDATE, DELETE

In [ ]:
# CREATE

runs_virat = [67, 54, 12, 34, 77, 89, 101, 92]

# The list can theoritically be infinitely sized - PYTHON HAS NO LIMIT ON LIST SIZE!

In [ ]:
runs_virat

[67, 54, 12, 34, 77, 89, 101, 92]

In [ ]:
type(runs_virat)

list

In [ ]:
# READ
# Values are accessible using indexes which start at 0

In [ ]:
runs_virat[0] # Runs scored in the first match - 0th index

67

In [ ]:
runs_virat[4] # Runs scored in the 5th match - 4th index

77

In [ ]:
len(runs_virat) # number of elements inside a list

8

In [ ]:
# HOW WOULD YOU GET THE RUNS SCORED IN THE LAST MATCH?

In [ ]:
runs_virat

[67, 54, 12, 34, 77, 89, 101, 92]

In [ ]:
n = len(runs_virat)

runs_virat[n - 1] # last element

92

In [ ]:
runs_virat[len(runs_virat) - 1]

92

In [ ]:
runs_virat[n - 2]

101

In [ ]:
# NEGATIVE INDEXING

In [ ]:
runs_virat[-1]

92

In [ ]:
runs_virat[-2]

101

In [ ]:
runs_virat[-n] # n here is the length as calculated above!

67

In [ ]:
runs_virat[100000] # These many elements are not present

IndexError: list index out of range

In [ ]:
runs_virat

[67, 54, 12, 34, 77, 89, 101, 92]

In [ ]:
# UPDATE

In [ ]:
# Adding elements to the list

In [ ]:
# 1. Append

runs_virat.append(123)

In [ ]:
runs_virat

[67, 54, 12, 34, 77, 89, 101, 92, 123]

In [ ]:
runs_virat = runs_virat + 45

TypeError: can only concatenate list (not "int") to list

In [ ]:
runs_virat = runs_virat + [45]

In [ ]:
runs_virat # The above operation is similar to a append operation

[67, 54, 12, 34, 77, 89, 101, 92, 123, 45]

In [ ]:
# 2. Insert

In [ ]:
# list.insert(index, value)

In [ ]:
runs_virat.insert(2, 55)

In [ ]:
runs_virat

[67, 54, 55, 12, 34, 77, 89, 101, 92, 123, 45]

In [ ]:
# 3. Extend

In [ ]:
runs_virat.extend([15, 19, 88])

In [ ]:
runs_virat

[67, 54, 55, 12, 34, 77, 89, 101, 92, 123, 45, 15, 19, 88]

In [ ]:
runs_virat += [1, 2, 3]

In [ ]:
runs_virat # The above operation is similar to the extend operation

[67, 54, 55, 12, 34, 77, 89, 101, 92, 123, 45, 15, 19, 88, 1, 2, 3, 1, 2, 3]

In [ ]:
'''
list.append(value) -> Add one value at the end of the list
    - list += [value]

list.insert(index, value) -> Add one value at the specified index in the list

list.extend([value1, value2, value3...]) -> Add multiple values at the end of the list
    - list += [value1, value2, value3...]
'''

'\nlist.append(value) -> Add one value at the end of the list\n    - list += [value]\n    \nlist.insert(index, value) -> Add one value at the specified index in the list\n\nlist.extend([value1, value2, value3...]) -> Add multiple values at the end of the list\n    - list += [value1, value2, value3...]\n'

In [ ]:
# Iteration on Lists

In [ ]:
a = "python"

for i in a:
    print(i)

p
y
t
h
o
n


In [ ]:
a = [10, 20, 30, 40, 50, 60]

In [ ]:
# TRAVERSAL METHOD 1 -> Directly traverse over the elements
for i in a:
    print(i)

10
20
30
40
50
60


In [ ]:
# TRAVERSAL METHOD 2 -> Using indexes
for i in range(len(a)):
    print(i, "->",a[i])

0 -> 10
1 -> 20
2 -> 30
3 -> 40
4 -> 50
5 -> 60


In [ ]:
# TRAVERSAL METHOD 3 -> Also possible using while loops

i = 0

while i < len(a):
    print(a[i])
    i += 1

10
20
30
40
50
60


In [ ]:
# LISTS ARE MUTABLE

In [ ]:
a = [1,2,3,4,5]

In [ ]:
a[0] = 10

In [ ]:
a

[10, 2, 3, 4, 5]

In [ ]:
# DELETE

In [ ]:
a = [10, 20, 30, 40, 50, 60]

In [ ]:
# 1. Delete by index -> list.pop(index)

In [ ]:
a.pop(2)

30

In [ ]:
a

[10, 20, 40, 50, 60]

In [ ]:
a.pop() # Automatically deletes the last element

60

In [ ]:
a

[10, 20, 40, 50]

In [ ]:
# 2. Delete by value -> list.remove(value)

In [ ]:
a = [10, 20, 30, 40, 50, 60]

In [ ]:
a.remove(40)

In [ ]:
a

[10, 20, 30, 50, 60]

In [ ]:
a.remove(100)

ValueError: list.remove(x): x not in list

In [ ]:
a = [10, 20, 30, 40, 50, 40, 60, 40]

In [ ]:
a.remove(40) # remove only deletes the first instance of a repeated value

In [ ]:
a

[10, 20, 30, 50, 40, 60, 40]

In [ ]:
50 in a # MEMBERSHIP OPERATOR

True

In [ ]:
100 in a

False

In [ ]:
# LISTS ARE HETEROGENEOUS

In [ ]:
a = ["bipin", 4.5, -9, True, [1,2,3,4]] # VALID LIST

In [ ]:
# LISTS SLICING

In [ ]:
'''
Question 1 -
Given a list of all runs,
print runs made in odd numbered (position) matches.
'''

'\nQuestion 1 -\nGiven a list of all runs,\nprint runs made in odd numbered (position) matches.\n'

In [ ]:
runs = [62, 85, 77, 10, 101, 122, 99, 81, 15, 36]

In [ ]:
# position n -> index n - 1

In [ ]:
# Odd numbered matches will occur on even indexes -
# 1st match - 0
# 3rd match - 2
# 5th match - 4
# .
# .
# .

In [ ]:
for i in range(len(runs)):
    if i % 2 == 0:
        print(runs[i], end = " ")

62 77 101 99 15 

In [ ]:
for i in range(0, len(runs), 2):
    print(runs[i], end = " ")

62 77 101 99 15 

In [ ]:
runs[:5] # start value is set at 0 by default

[62, 85, 77, 10, 101]

In [ ]:
runs[6:] # end value is set at len(list) by default

[99, 81, 15, 36]

In [ ]:
runs[:]

[62, 85, 77, 10, 101, 122, 99, 81, 15, 36]

In [ ]:
runs[4000]

IndexError: list index out of range

In [ ]:
runs[4000:5000]

[]

In [ ]:
runs[2:4000]

[77, 10, 101, 122, 99, 81, 15, 36]

In [ ]:
# Runs in even numbered matches -> Odd Indexes
runs[1::2]

[85, 10, 122, 81, 36]

In [ ]:
runs[5:0]

[]

In [ ]:
runs[5:0:-1]

[122, 101, 10, 77, 85]

In [ ]:
'''
result = []

for i in range(-1, -len(runs) - 1, -1):
    result.append(runs[i])
'''

'\nresult = []\n\nfor i in range(-1, -len(runs) - 1, -1):\n    result.append(runs[i])\n'

In [ ]:
'''
Default values -
1. When jump / step size is positive -
    - start = 0
    - end = len(list)
2. When jump / step size is negative -
    - start = -1
    - end = -len(runs) - 1
'''

'\nDefault values - \n1. When jump / step size is positive - \n    - start = 0\n    - end = len(list)\n2. When jump / step size is negative - \n    - start = -1\n    - end = -len(runs) - 1\n'

In [ ]:
#TO REVERSE ANY LIST IN PYTHON
runs[::-1]

[36, 15, 81, 99, 122, 101, 10, 77, 85, 62]

In [ ]:
squares = [] # 1. Creating a result list

for i in range(1, 11): # 2. Writing a loop
    squares.append(i ** 2) #3.  Appending a value

In [ ]:
squares

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [ ]:
squares_amazing = [ i**2 for i in range(1, 11) ]
# variable_name = [ value_to_be_apended for i in range(start, end) ]

In [ ]:
squares_amazing

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [ ]:
table = []

for i in range(1, 11):
    row = []

    for j in range(1, 11):
        row.append(i*j)

    table.append(row)

In [ ]:
table

[[1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
 [2, 4, 6, 8, 10, 12, 14, 16, 18, 20],
 [3, 6, 9, 12, 15, 18, 21, 24, 27, 30],
 [4, 8, 12, 16, 20, 24, 28, 32, 36, 40],
 [5, 10, 15, 20, 25, 30, 35, 40, 45, 50],
 [6, 12, 18, 24, 30, 36, 42, 48, 54, 60],
 [7, 14, 21, 28, 35, 42, 49, 56, 63, 70],
 [8, 16, 24, 32, 40, 48, 56, 64, 72, 80],
 [9, 18, 27, 36, 45, 54, 63, 72, 81, 90],
 [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]]

In [ ]:
t = [ [ i*j for j in range(1,11) ] for i in range(1,11) ]

In [ ]:
t

[[1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
 [2, 4, 6, 8, 10, 12, 14, 16, 18, 20],
 [3, 6, 9, 12, 15, 18, 21, 24, 27, 30],
 [4, 8, 12, 16, 20, 24, 28, 32, 36, 40],
 [5, 10, 15, 20, 25, 30, 35, 40, 45, 50],
 [6, 12, 18, 24, 30, 36, 42, 48, 54, 60],
 [7, 14, 21, 28, 35, 42, 49, 56, 63, 70],
 [8, 16, 24, 32, 40, 48, 56, 64, 72, 80],
 [9, 18, 27, 36, 45, 54, 63, 72, 81, 90],
 [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]]

#**`Python Strings`**

Strings are sequences of characters in Python, and they have various methods for manipulation.

**Glossary of string methods**

1. __split__ -> `"this is a string".split(" ")` -> string to a list

2. __join__ -> `" ".join(["list", "of", "strings"])` -> list of strings to a string

3. __replace__ -> `"this is a string".replace(" ", "_")` -> replaces all occurences of the first character by the second character. __IMP__ - _CREATES A NEW STRING_

4. __find__ -> `"this is a string".find(" a ")` -> Finds the exact sequence or substring in the original string and returns the starting index of that substring.

5. __count__ -> `"this is a string".count("a")` -> Counts the number of times a character or substring is present within a string

6. __isdigit__ -> True if content inside the string is only digits.
7. __isalpha__ -> True if content inside the string is only alphabets.
8. __islower__ / __isupper__ -> True if content inside the string is all lower case / all upper case
9. __isspace__ -> True if the character is a space
10. __lower__ / __upper__ -> Converts a string to lowercase / uppercase

__Special Mention__
- `"random" in "this is random!"` -> MEMBERSHIP OPERATOR - True if operand1 is present inside operand2.

In [ ]:
a = "double quotes!"

In [ ]:
b = 'single quotes!'

In [ ]:
c = '''triple quotes! -> DOCSTRINGS'''

In [ ]:
d = """triple double quotes! -> DOCSTRINGS"""

In [ ]:
type(d)

str

In [ ]:
randoom = "Whatever we write, UPPERCASE, lowercase, 12345678908765432@#$%^&*(*&^%$#@#$%^)"

In [ ]:
randoom

'Whatever we write, UPPERCASE, lowercase, 12345678908765432@#$%^&*(*&^%$#@#$%^)'

In [ ]:
random = "a' # Mixed quotations are not allowed

SyntaxError: EOL while scanning string literal (4111874266.py, line 1)

In [ ]:
# What if your string itself contains quotes inside it

y = "something "random" which is highlight!"

# If you want to use double quotes inside a string
# -> Use single quotes to form the string!
# And vice versa

SyntaxError: invalid syntax (3829515004.py, line 3)

In [ ]:
y = 'something "random" which is highlighted!'

In [ ]:
y

'something "random" which is highlighted!'

In [ ]:
z = "something 'random' which is highlighted!"

In [ ]:
z

"something 'random' which is highlighted!"

In [ ]:
a = 'can't' # THIS HAS A LOT OF ISSUES!

SyntaxError: invalid syntax (3506376244.py, line 1)

In [ ]:
a = 'can\'t' # Escape character

In [ ]:
a

"can't"

In [ ]:
random = "this string is \"random\" which can't work properly without \"escape characters!\""

In [ ]:
random

'this string is "random" which can\'t work properly without "escape characters!"'

In [ ]:
"zebra" > "aaaaaahgfshjfgjhchsdudadhgafgdadaajdghdja"
# LEXICOGRAPHICAL - Order in a dictionary

True

In [ ]:
"abc" > "agh"

False

In [ ]:
# This comparison is enabled via mapping in programming!

In [ ]:
"a" > "A" # THIS IS BECAUSE OF ASCII!

True

In [ ]:
"zebra" > "ZEBRA"

True

In [ ]:
"random1" + "random2" # Concatenation

'random1random2'

In [ ]:
"random" * 2 * 3

'randomrandomrandomrandomrandomrandom'

In [ ]:
ord("a") # Corresponding ASCII (mapping) value

97

In [ ]:
ord("z")

122

In [ ]:
ord("A")

65

In [ ]:
ord("%")

37

In [ ]:
ord("0")

48

In [ ]:
chr(97) # Corresponding character acc to mapping

'a'

In [ ]:
a = "a string with a few words separated by space."

In [ ]:
a.split(" ") # A delimiter is provided as an argument

['a', 'string', 'with', 'a', 'few', 'words', 'separated', 'by', 'space.']

In [ ]:
a.split("s")

['a ', 'tring with a few word', ' ', 'eparated by ', 'pace.']

In [ ]:
a = "this+is+random"

In [ ]:
a.split("+")

['this', 'is', 'random']

In [ ]:
a = "thisRANDOMisRANDOMveryRANDOMrandom!"

In [ ]:
a.split("RANDOM")

['this', 'is', 'very', 'random!']

In [ ]:
'''
Using split,
we split a string into a list of strings, delimited at a character(s).

Using join,
we join a list of strings into one string using a joining character(s).
'''

'\nUsing split,\nwe split a string into a list of strings, delimited at a character(s).\n\nUsing join,\nwe join a list of strings into one string using a joining character(s).\n'

In [ ]:
a = ['a', 'string', 'with', 'a', 'few', 'words', 'separated', 'by', 'space.']

In [ ]:
a

['a', 'string', 'with', 'a', 'few', 'words', 'separated', 'by', 'space.']

In [ ]:
" ".join(a) # "delimter".join(list_of_strings)

'a string with a few words separated by space.'

In [ ]:
"_+_".join(a)

'a_+_string_+_with_+_a_+_few_+_words_+_separated_+_by_+_space.'

In [ ]:
a = ["random", 23, 5.6, True]

In [ ]:
a = "this is a random string!"

In [ ]:
a.find("is") # It would return the first index of the substring in the string

2

In [ ]:
a.find("random")

10

In [ ]:
a[a.find("random"):a.find("random")+len("random")]

'random'

In [ ]:
a[:a.find("random")] + a[a.find("random") + len("random") + 1:]

'this is a string!'

In [ ]:
a.find("not present!")
# Returns -1 if the substring can't be found

-1

In [ ]:
# 4. REPLACE

In [ ]:
a = "this is a random string with very random random words arranged in a random order which is random!"

In [ ]:
a

'this is a random string with very random random words arranged in a random order which is random!'

In [ ]:
a.replace("random", "SUPERB")

'this is a SUPERB string with very SUPERB SUPERB words arranged in a SUPERB order which is SUPERB!'

In [ ]:
a # Does not modify the original string

'this is a random string with very random random words arranged in a random order which is random!'

In [ ]:
a_replaced = a.replace("random", "SUPERB")

In [ ]:
a_replaced

'this is a SUPERB string with very SUPERB SUPERB words arranged in a SUPERB order which is SUPERB!'

In [ ]:
# 5. MEMBERSHIP OPERATOR

In [ ]:
a = "this is a random string!"

In [ ]:
"random" in a

True

In [ ]:
"RANDOM" in a

False

In [ ]:
"something else" in a

False

In [ ]:
"random string" in a

True

In [ ]:
"56.7".isdigit()
# This string contains "." also which is not a number

False

In [ ]:
"smallCAPITAL".isalpha()

True

In [ ]:
"sdhfgs23456789".isalnum()
# Returns True if the string only contains alphabets / numbers

True

In [ ]:
"a".isupper()

False

In [ ]:
"A".islower()

False

In [ ]:
"This IS A random STRing!".upper()

'THIS IS A RANDOM STRING!'

In [ ]:
"This IS A random STRing!".lower()

'this is a random string!'

In [ ]:
" ".isspace()

True

#**`Python Tuples`**

Tuples are similar to lists but have a key distinction: they are immutable. This means that once you create a tuple, you cannot change its elements. Tuples are created using parentheses. Here's more detail:


| Aspect              | Description                                                       | Example                                        |
|---------------------|-------------------------------------------------------------------|------------------------------------------------|
| **Creation**        | Tuples are defined using parentheses and can contain various data types. | `my_tuple = (1, "apple", 3.14)` |
| **Accessing Elements** | Elements in a tuple can be accessed using indexing, similar to lists. | `print(my_tuple[0])` # Access the first element |
| **Immutability**    | Tuples are immutable, meaning their elements cannot be modified or changed after creation. | Once created, the tuple cannot be modified. |
| **Tuple Methods**   | Tuples have methods like `count()` and `index()` for counting occurrences of an element and finding the index of an element, respectively. | `my_tuple = (1, 2, 2, 3, 4)`<br>`count = my_tuple.count(2)` # Counts how many times 2 appears<br>`index = my_tuple.index(3)` # Finds the index of 3 |


In [ ]:
a = [1, 2, 3, 4]

In [ ]:
a.append(5)

In [ ]:
a

[1, 2, 3, 4, 5]

In [ ]:
# The size of the list is not fixed

In [ ]:
a[0] = 1.0

In [ ]:
a

[1.0, 2, 3, 4, 5]

In [ ]:
# List elements are not fixed either

In [ ]:
# DEFINING A TUPLE - Use paranthesis
a = (1, 2, 3, 4)

In [ ]:
a.append(5)

AttributeError: 'tuple' object has no attribute 'append'

In [ ]:
a.remove(4)

AttributeError: 'tuple' object has no attribute 'remove'

In [ ]:
a[0] = 100

TypeError: 'tuple' object does not support item assignment

In [ ]:
a = tuple(range(1, 11))

In [ ]:
a

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

In [ ]:
a[6]

7

In [ ]:
a[-1]

10

In [ ]:
a[3:7]

(4, 5, 6, 7)

In [ ]:
for i in a:
    print(i, end = " ")

1 2 3 4 5 6 7 8 9 10 

In [ ]:
a

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

In [ ]:
type(a)

tuple

In [ ]:
l = []

In [ ]:
type(l)

list

In [ ]:
t = ()

In [ ]:
type(t)

tuple

In [ ]:
l = [1]

In [ ]:
type(l)

list

In [ ]:
t = (1)

In [ ]:
type(t)

int

In [ ]:
t = ((45 + 56) * 2) // 3 # paranthesis are used to build mathematical equations

In [ ]:
t = (1,) # This extra comma tells python that it is a tuple

In [ ]:
type(t)

tuple

In [ ]:
t

(1,)

In [ ]:
for i in t:
    print(i)

1


In [ ]:
a = [1, 2, 3, 4, 5]

In [ ]:
tuple(a)

(1, 2, 3, 4, 5)

In [ ]:
t1 = (1,2,3)
t2 = (4,5,6)

In [ ]:
id(t1)

140628798051136

In [ ]:
id(t2)

140628239723136

In [ ]:
t3 = t1 + t2

In [ ]:
t3

(1, 2, 3, 4, 5, 6)

In [ ]:
id(t3)

140628798017152

In [ ]:
'''
Question 1 -
What will be the output of the following?
(4) + (1,2,3)
'''

'\nQuestion 1 - \nWhat will be the output of the following?\n(4) + (1,2,3)\n'

In [ ]:
(4) + (1,2,3)

TypeError: unsupported operand type(s) for +: 'int' and 'tuple'

In [ ]:
'''
Question 2 -
What will be the output of the following?
type(("random?"))
'''

'\nQuestion 2 - \nWhat will be the output of the following?\ntype(("random?"))\n'

In [ ]:
type(("random?"))

str

In [ ]:
type(("random?",))

tuple

In [ ]:
# PACKING AND UNPACKING!

In [ ]:
a = (1, 2, 3)

In [ ]:
x, y, z = a # UNPACKING tuple a into variables x, y and z

In [ ]:
x

1

In [ ]:
y

2

In [ ]:
z

3

In [ ]:
x, y = a

# THERE IS A SOLUTION TO THIS WHICH WE WILL LEARN IN "ADVANCE PYTHON"

ValueError: too many values to unpack (expected 2)

In [ ]:
w, x, y, z = a

ValueError: not enough values to unpack (expected 4, got 3)

In [ ]:
t = ("string", 5, 5.6, True, (1,2,3), [1,2,3]) # HETEROGENEOUS!

In [ ]:
t

('string', 5, 5.6, True, (1, 2, 3), [1, 2, 3])

#**`Python Dictionaries`**

#**`Python Sets`**

Sets are collections of unique elements. They are defined using curly braces or the set() constructor. Here's more about sets:


| Aspect              | Description                                                       | Example                                        |
|---------------------|-------------------------------------------------------------------|------------------------------------------------|
| **Creation**        | Sets can be defined using curly braces or the `set()` constructor.  | `my_set = {1, 2, 3}`<br>`another_set = set([3, 4, 5])` |
| **Uniqueness**      | Sets do not allow duplicate elements. If you try to add an element that already exists in the set, it will not be added again. | `my_set = {1, 2, 2, 3}` results in `my_set` being `{1, 2, 3}` |
| **Set Operations**  | Sets support various set operations like union, intersection, difference, and symmetric difference. These operations allow you to manipulate sets efficiently. | `set1 = {1, 2, 3}`<br>`set2 = {3, 4, 5}`<br>`union_set = set1.union(set2)`<br>`intersection_set = set1.intersection(set2)`<br>`difference_set = set1.difference(set2)` |


In [ ]:
# SETS ARE A COLLECTION OF UNIQUE VALUES

In [ ]:
a = {1,2,2,2,2,2,3,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,6,7,7,7,7,8,8,9,9}

In [ ]:
a

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [ ]:
a[0]

TypeError: 'set' object is not subscriptable

In [ ]:
a = {}

In [ ]:
type(a)

dict

In [ ]:
a = set() # ONLY WAY TO CREATE AN EMPTY SET!

In [ ]:
a = {4, 3, 1, 2, 7}

In [ ]:
a

{1, 2, 3, 4, 7}

In [ ]:
a = {1, 2, 3}

In [ ]:
a.add(5)

In [ ]:
a

{1, 2, 3, 5}

In [ ]:
a.remove(3)

In [ ]:
a

{1, 2, 5}

In [ ]:
a.update([1,2,3,4,4,5,5,6,6,8,9,0,10])

In [ ]:
a

{0, 1, 2, 3, 4, 5, 6, 8, 9, 10}

In [ ]:
l = [1, 2, 3, 4, 5]

In [ ]:
t = tuple(l)

In [ ]:
t

(1, 2, 3, 4, 5)

In [ ]:
s = set(l)

In [ ]:
s

{1, 2, 3, 4, 5}

In [ ]:
t1 = tuple(s)

In [ ]:
t1

(1, 2, 3, 4, 5)

In [ ]:
'''
Question 6 -
Take a string as input.
Print the number of unique characters present inside that string.

"aabbcccd" -> 4
'''

'\nQuestion 6 - \nTake a string as input.\nPrint the number of unique characters present inside that string.\n\n"aabbcccd" -> 4\n'

In [ ]:
a = "this is a random string with very random randomness!"

In [ ]:
set(a)

{' ',
 '!',
 'a',
 'd',
 'e',
 'g',
 'h',
 'i',
 'm',
 'n',
 'o',
 'r',
 's',
 't',
 'v',
 'w',
 'y'}

In [ ]:
len(set(a)) # COMPLETE SOLUTION!

17

In [ ]:
s1 = {2, 3, 4}
s2 = {1, 2, 4, 5, 6}

In [ ]:
s1.intersection(s2)

{2, 4}

In [ ]:
s1 & s2

{2, 4}

In [ ]:
s1.union(s2)

{1, 2, 3, 4, 5, 6}

In [ ]:
s1 | s2

{1, 2, 3, 4, 5, 6}

In [ ]:
s1.difference(s2)

{3}

In [ ]:
s1 - s2

{3}

In [ ]:
s1.symmetric_difference(s2)

{1, 3, 5, 6}

In [ ]:
s1 ^ s2

{1, 3, 5, 6}

In [ ]:
# Sets can't contain mutable datatypes
# Sets can have everything but -> lists, dictionaries, sets

In [ ]:
a = {1, "string", 4.5, True, (1,2,3)}

In [ ]:
type(a)

set

In [ ]:
a = {1, "string", 4.5, True, [1,2,3]}

TypeError: unhashable type: 'list'

Dictionaries are collections of key-value pairs. They are used for efficient data retrieval and storage. Here are some details about dictionaries:

| Aspect              | Description                                                       | Example                                        |
|---------------------|-------------------------------------------------------------------|------------------------------------------------|
| **Creation**        | Dictionaries are created using curly braces and key-value pairs.  | `my_dict = {"name": "Alice", "age": 30, "city": "New York"}` |
| **Key-Value Pairs** | Each element in a dictionary consists of a key and its associated value. Keys must be unique within the dictionary. | `"name": "Alice"`<br>`"age": 30`<br>`"city": "New York"` |
| **Accessing Values** | You can access values in a dictionary by using the keys. | `name = my_dict["name"]` |
| **Dictionary Methods** | Dictionaries have methods like `keys()`, `values()`, and `items()` for working with the keys, values, and key-value pairs, respectively. | `keys = my_dict.keys()`<br>`values = my_dict.values()`<br>`items = my_dict.items()` |

In [ ]:
s = {"bipin", 12, 34, 56, "random"}

In [ ]:
s

{12, 34, 56, 'bipin', 'random'}

In [ ]:
d = {
    "key1" : "value1",
    "key2" : "value2",
    "key3" : "value3"
}

# For keys, we will only use strings for a while - SECOND HALF OF LECTURE
# For values, we can use anything!

In [ ]:
# Approach 1 -> List of tuples
state_wise_data = [("Delhi", 15), ("Haryana", 35), ("UP", 100)]

In [ ]:
state_wise_data[0][1]

15

In [ ]:
# Approach 2 -> Dictionary
state_wise_data = {
    "Delhi" : 15,
    "Haryana" : 35,
    "UP" : 100
}

In [ ]:
state_wise_data[0]

# Similar to sets, dictionaries are not subscriptable via numerical indexes
# They can be indexed via keys!

KeyError: 0

In [ ]:
state_wise_data["Delhi"]

15

In [ ]:
state_wise_data["Haryana"]

35

In [ ]:
state_wise_data.Delhi

AttributeError: 'dict' object has no attribute 'Delhi'

In [ ]:
a = {}

In [ ]:
type(a)

dict

In [ ]:
# KEYS ARE ALWAYS UNIQUE

In [ ]:
a = {
    "Delhi" : 15,
    "Delhi" : 10,
    "Delhi" : 30,
    "Haryana" : 36,
    "UP" : 100,
    "Delhi" : 20
}

In [ ]:
a

{'Delhi': 20, 'Haryana': 36, 'UP': 100}

In [ ]:
# Keys will never be repeated
# The last or latest value that is provided will be kept!

In [ ]:
# ADDING MORE KEY VALUE PAIRS

In [ ]:
a["Maharashtra"] = 80

In [ ]:
a

{'Delhi': 20, 'Haryana': 36, 'UP': 100, 'Maharashtra': 80}

In [ ]:
a["Delhi"]

20

In [ ]:
a["delhi"] = 35

In [ ]:
a

{'Delhi': 20, 'Haryana': 36, 'UP': 100, 'Maharashtra': 80, 'delhi': 35}

In [ ]:
Avenger = {
    "name" : "Thor",
    "age" : 1500,
    "weapon" : ["Stormbreaker, Mjolnir"]
}

In [ ]:
Avenger.update({
    "name" : "Thor Odinson",
    "weapon" : ["Stormbreaker"],
    "strongest?" : True,
    "eyes" : 1
})

In [ ]:
Avenger

{'name': 'Thor Odinson',
 'age': 1500,
 'weapon': ['Stormbreaker'],
 'strongest?': True,
 'eyes': 1}

In [ ]:
random = {
    "a" : 1,
    "b" : 2,
    "c" : 3,
    "d" : None
}

In [ ]:
random["b"]

2

In [ ]:
random["e"]

KeyError: 'e'

In [ ]:
# GET FUNCTION

In [ ]:
result = random.get("e")

In [ ]:
print(result) # Returns None instead of throwing an error!

None


In [ ]:
random.get("b")

2

In [ ]:
random.get("e", "NOT FOUND!")

'NOT FOUND!'

In [ ]:
random.get("b", "NOT FOUND!")

2

In [ ]:
random

{'a': 1, 'b': 2, 'c': 3, 'd': None}

In [ ]:
random.pop("d") # provide the key as argument

In [ ]:
random

{'a': 1, 'b': 2, 'c': 3}

In [ ]:
random.pop("a")

1

In [ ]:
random

{'b': 2, 'c': 3}

In [ ]:
random.pop()

TypeError: pop expected at least 1 argument, got 0

In [ ]:
random = {1,2,3} | {3,4,5}

In [ ]:
random

{1, 2, 3, 4, 5}

In [ ]:
result = A | B # Union in dictionaries

In [ ]:
result

{'a': 4, 'b': 1, 'c': 6, 'd': 20}

In [ ]:
result = { **A, **B } # Spread Operator

In [ ]:
result

{'a': 4, 'b': 1, 'c': 6, 'd': 20}

In [ ]:
# ITERATION IN A DICTIONARY

In [ ]:
a = Avenger

In [ ]:
a

{'name': 'Thor Odinson',
 'age': 1500,
 'weapon': ['Stormbreaker'],
 'strongest?': True,
 'eyes': 1}

In [ ]:
for i in a: # by default it iterates over the keys
    print(i)

name
age
weapon
strongest?
eyes


In [ ]:
for i in a: # to go over the values
    print(a[i])

Thor Odinson
1500
['Stormbreaker']
True
1


In [ ]:
for i in a:
    print(f"{i} -> {a[i]}")

name -> Thor Odinson
age -> 1500
weapon -> ['Stormbreaker']
strongest? -> True
eyes -> 1


In [ ]:
list(a.keys())

['name', 'age', 'weapon', 'strongest?', 'eyes']

In [ ]:
list(a.values())

['Thor Odinson', 1500, ['Stormbreaker'], True, 1]

In [ ]:
list(a.items())

[('name', 'Thor Odinson'),
 ('age', 1500),
 ('weapon', ['Stormbreaker']),
 ('strongest?', True),
 ('eyes', 1)]

In [ ]:
for i in a.keys():
    print(i)

name
age
weapon
strongest?
eyes


In [ ]:
for i in a.values():
    print(i)

Thor Odinson
1500
['Stormbreaker']
True
1


In [ ]:
for i in a.items():
    print(i)

('name', 'Thor Odinson')
('age', 1500)
('weapon', ['Stormbreaker'])
('strongest?', True)
('eyes', 1)


In [ ]:
for key, value in a.items():
    print(f"{key} -> {value}")

name -> Thor Odinson
age -> 1500
weapon -> ['Stormbreaker']
strongest? -> True
eyes -> 1


In [ ]:
# random(a) / a.random()

In [ ]:
"strongest?" in a

True

In [ ]:
"RANDOM" in a

False

In [ ]:
'''
HETEROGENEITY ->
    - Keys -> Any immutable data structure (Everything other than lists, sets and dictionaries)
    - Values -> Any Python Object
'''

'\nHETEROGENEITY ->\n    - Keys -> Any immutable data structure (Everything other than lists, sets and dictionaries)\n    - Values -> Any Python Object\n'

#**`Linked List`**

**Node Class**


A node is the basic building block of a linked list. It contains data and a reference to the next node in the sequence.

In [ ]:
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

**Linked List Class:**

The linked list class contains methods for managing the linked list, such as appending, prepending, deleting nodes, and displaying the list.

**1. Initialization and is_empty Method:**  

- **init Method**: Initializes an empty linked list with a head set to None.
- **is_empty Method**: Checks if the linked list is empty. Returns True if the head is None, indicating an empty list.


In [ ]:
class LinkedList:
    def __init__(self):
        self.head = None

    def is_empty(self):
        return self.head is None


**2. Append Method:**  

- **append Method**: Adds a new node with the given data to the end of the linked list.
- If the list is empty (self.head is None), sets the head to the new node.
- Otherwise, traverses the list to find the last node and appends the new node to it.


In [ ]:
    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last_node = self.head
        while last_node.next:
            last_node = last_node.next
        last_node.next = new_node


**3. prepend Method:**  

- **prepend Method**: Adds a new node with the given data to the beginning of the linked list.
- Creates a new node with the given data.
- Sets the next reference of the new node to the current head.
- Updates the head to the new node.


In [ ]:
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node


**4. delete_node Method:**  

- **delete_node Method**: Deletes the node with the specified data from the linked list.
- Handles the case where the node to be deleted is the head.
- Traverses the list to find the node with the specified data.
- Updates the references to bypass the node to be deleted.

In [ ]:
    def delete_node(self, key):
        current_node = self.head
        if current_node and current_node.data == key:
            self.head = current_node.next
            current_node = None
            return
        prev_node = None
        while current_node and current_node.data != key:
            prev_node = current_node
            current_node = current_node.next
        if current_node is None:
            return
        prev_node.next = current_node.next
        current_node = None


**5. display Method:**  

- **display Method**: Displays the elements of the linked list.
- Starts from the head and iterates through the list, printing each node's data followed by an arrow (->).
- Prints "None" at the end to indicate the end of the list.






In [ ]:
    def display(self):
        current_node = self.head
        while current_node:
            print(current_node.data, end=" -> ")
            current_node = current_node.next
        print("None")
